### PR TITLE
Tracker component knows about Agents and Environments + Run publishing and downloading

### PR DESCRIPTION
**Pull request by nickjalbert**
# Update 11/29/2021

This PR now includes implementation of the full loop:  running an agent, uploading the run, downloading the component spec and backing data from that run, and (stochastically) replaying that run.

This is deployed on [Heroku](https://aos-web.herokuapp.com/), but you can also run it against the development server (note, the migrations have changed so you may have to dump your postgres database if you're running locally):

Here's a demo (including a bunch of setup that you may or may not need depending on the state of your local repo):

```bash
# Get my branch
git clone https://github.com/nickjalbert/agentos.git
cd agentos
git checkout nj_cli_run_publish_2
virtualenv -p /usr/bin/python3.9 env
source env/bin/activate


# Setup and run a web server
cd web
pip install -r requirements.txt
sudo -u postgres psql   # Time to dump and re-create the DB
drop database aos_web; drop user aos_web_user; create database aos_web; create user aos_web_user with encrypted password 'aabbccdd'; grant all privileges
./manage.py migrate
./manage.py create_default_admin
./manage.py runserver

# New tab, let's benchmark an agent and upload the run to the local server
cd agentos
echo "USE_LOCAL_SERVER=True" >> .env
source env/bin/activate
pip install -e .
cd example_agents/sb3_agent/
pip install -r requirements.txt  # install SB3 requirements
rm -rf mlruns  # kill any previous state
agentos run agent --entry-point learn -P total_timesteps=5000  # Train our agent
agentos run agent  # Benchmark our agent
agentos run tracker # Publish our benchmark run

# Navigate to http://localhost:8000 and confirm your run appears there

# Now repro the run
cd ../..
mkdir repro_demo
cd repro_demo
agentos get-run 1
ls  # Note that we've create a spec file and some other useful things in the dir
agentos run agent  # stochastically reproduce your benchmarked run
```

One more small thing I've added is that you can define a default entry point by create a `DEFAULT_ENTRY_POINT` static variable on a class (example on Tracker [here](https://github.com/agentos-project/agentos/pull/204/files#diff-b52f4d2f4a5fb71006dd6044a42f93b9a569335472fda775bafcdb3522a24c24R39)).

# Original

This PR implements the second bullet point from our "PCS+AOS brainstorming" email (Nov 15, 2021), namely:

> * Tracker component is the interface with the leaderboard
>    * with self.tracker.evaluate_run takes two new args, the agent component and the env component
>    * whenever we start a run in PCS, we attempt to pin the component we're running
>      * If we can pin, great we can upload the run later to the leaderboard
>      * If we can't pin, we can issue a warning but still execute the run anyway

As discussed in our standup today, the Tracker component will simply record a (user-specifiable, but with sane defaults) name for the Agent component and the Environment component and confirm those exists in the executing component tree.  This lets us specify the Agent and Environment when we push a component spec and run data to the server.

## Demo

```bash
cd example_agents/sb3_agent/

# Perform a benchmark run
agentos run agent

# Prints all information about the benchmark run; doesn't upload yet
agentos run tracker --entry-point publish
```

On my box, `tracer.publish`  prints out something like:

```
{'agent_exists': 'True',
 'agent_name': 'agent',
 'artifact_paths': [PosixPath('/local/path/to/the/saved/model/ppo.zip')],
 'component_spec': {'components': {'agent==5b0c5ad5af68ab753226f55e40601b621de7c08b': {'class_name': 'SB3PPOAgent',
                                                                                       'dependencies': {'environment': 'environment==5b0c5ad5af68ab753226f55e40601b621de7c08b',
                                                                                                        'tracker': 'tracker==5b0c5ad5af68ab753226f55e40601b621de7c08b'},
                                                                                       'file_path': 'example_agents/sb3_agent/agent.py',
                                                                                       'repo': 'sb3_agent_dir'},
                                   'environment==5b0c5ad5af68ab753226f55e40601b621de7c08b': {'class_name': 'CartPole',
                                                                                             'dependencies': {},
                                                                                             'file_path': 'example_agents/sb3_agent/environment.py',
                                                                                             'repo': 'sb3_agent_dir'},
                                   'tracker==5b0c5ad5af68ab753226f55e40601b621de7c08b': {'class_name': 'SB3Tracker',
                                                                                         'dependencies': {},
                                                                                         'file_path': 'example_agents/sb3_agent/tracker.py',
                                                                                         'repo': 'sb3_agent_dir'}},
                    'repos': {'sb3_agent_dir': {'type': 'github',
                                                'url': 'https://github.com/nickjalbert/agentos.git'}}},
 'entry_point': 'evaluate',
 'environment_exists': 'True',
 'environment_name': 'environment',
 'metrics': {'episode_count': 10.0,
             'max_reward': 501.0,
             'mean_reward': 501.0,
             'median_reward': 501.0,
             'min_reward': 501.0,
             'step_count': 5010.0,
             'training_episode_count': 351.0,
             'training_step_count': 51200.0},
 'parameters': {'agent': {'evaluate': {}}},
 'spec_is_frozen': 'True'}
```

Note that the spec is frozen even though we run against an unfrozen spec (if we are unable to freeze it, we print a warning and record that this run is not publishable).  

Of the data keys associated with the run above, the runtime/PCS is responsible for:
* `component_spec`
* `entry_point`
* `parameters`
* `spec_is_frozen`

and Tracker is responsible for:
* `agent_name`
* `agent_exists`
* `artifact_paths`
* `environment_name`
* `environment_exists`
* `metrics`

This is roughly the data blob I envision pushing to the server to publish a run.

### CODE REVIEW

**Comment by nickjalbert** [comment]
@andyk this PR now contains all unmerged work from my local branches.  PR description has been updated!

**Comment by nickjalbert** [comment]
@andyk I've fixed the bug we found during the demo today where different environments with the same name+version were getting grouped together.  Now, the server will complain if you try to upload different Components with the same name.

For example:

```bash

# Let's run and publish an Acme DQN run
cd example_agents/acme_dqn/

agentos run agent --param-file parameters.yaml
    Adding network to agent
    Adding environment to network
    ...

agentos run tracker
    Calling tracker.__init__(**{})
    Calling tracker.publish(**{})
    Pushed Run 7 on http://localhost:8000

# Now let's run and publish an SB3 PPO agent run
cd ../sb3_agent/

agentos run agent
    Adding environment to agent
    Calling environment.__init__(**{})
    ...

# This publish fails because of a name conflict between the two different environment components
agentos run tracker
    ...
    Exception: Component with name agent==877c2d1103bc3135eec83513c8daa0e6cc4f823b (id: 21) already exists and differs from uploaded spec. Try renaming your agent==877c2d1103bc3135eec83513c8daa0e6cc4f823b Component.

File: agentos/component.py [hunk 1]
```diff
@@ -173,6 +195,13 @@ def _parse_components(components_spec: Dict, repos: Dict) -> Dict:
 
         return components
 
+    def get_default_entry_point(self):
+        try:
+            entry_point = self._managed_cls.DEFAULT_ENTRY_POINT
+        except AttributeError:
+            entry_point = "evaluate"
```
Comment (nickjalbert) [inline agentos/component.py:202]: Change to `run` by default.

File: agentos/tracker.py [hunk 2]
```diff
@@ -31,81 +36,116 @@ class AgentTracker:
     A Component used to track Agent training and evaluation runs.
     """
 
+    DEFAULT_ENTRY_POINT = "publish"
```
Comment (nickjalbert) [inline agentos/tracker.py:39]: Maybe change default to `status` command.  Bigger chunk of work: moving the publish/get functionality into lower-layer PCS.

File: agentos/tracker.py [hunk 3]
```diff
@@ -221,29 +261,46 @@ def add_episode_data(self, steps: int, reward: float):
...
+        tracker: AgentTracker,
+        run_type: str,
+        agent_name: Optional[str],
+        environment_name: Optional[str],
+    ):
         self.tracker = tracker
+        self.run_type = run_type
+        self.agent_name = agent_name or "agent"
+        self.environment_name = environment_name or "environment"
+        self._check_component_exists_in_run("agent")
```
Comment (nickjalbert) [inline agentos/tracker.py:280]: should be `self.agent_name` and `self.environment_name`

File: agentos/web.py [hunk 4]
```diff
@@ -1,17 +1,132 @@
...
+        files = {"tarball": open(tar_gz_path, "rb")}
+        url = f"{AOS_WEB_API_ROOT}/runs/{run_id}/upload_artifact/"
+        response = requests.post(url, files=files)
+        result = json.loads(response.content)
+        return result
+    finally:
+        shutil.rmtree(tmp_dir_path)
+
+
+def get_run(run_id: int) -> None:
```
Comment (nickjalbert) [inline agentos/web.py:72]: Future work: re-running without dumping a bunch of files in your cwd

**Review by andyk — APPROVED** [review]
Thanks for the PR walk-through via Zoom just now! This look great. After you fix the small bug in Tracker (`_check_component_exists_in_run` should be called with `self.agent_name` and `self.environment_name`), this LGTM to merge!

**Comment by andyk** [comment]
Below are notes from the walkthrough of this PR that you just gave me in our working session.

After we merge this PR, we discussed two high level things to refactor in a follow-on PR:
1. Make Run first class part of PCS
2. Support all functionality from inside Python interpreter (via core class APIs) & wrap that w/ CLI

And below is a more detailed list of things that would be involved in the two high-level items above:
- unify MLflow RunID and web.py RunID
- Make Run a first class abstraction in PCS (as a peer w/ Component, Repo, ParameterSet)
    - For now, probably use mlflow.Run
- Make agentos.RunManager a first class abstraction in PCS w/ publish functionality
- Where does a component developer annotate default_entry_point
- Rename Tracker --> AgentRunManager; add "status" fn and default entry-point to be it
- move web.py functionality into component.RunManager
- Allow python interpreter (Py API only) publishing Component, Run
- Make local runs of local git-repo (i.e., not published in github) do the auto-freeze
    - Auto capture diffs between current code and most recent commit

**Comment by nickjalbert** [comment]
Great!  Bug fixed; will merge when build is green.

**Pull request merged**